### ISSUE TITLE
odd bug with handle dup

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6410]
Migrated from [rt.perl.org#21780](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=21780) (status was 'resolved')

Searchable as RT21780$

**Comment by p5pRT** [comment]
### From belg4mit@mit.edu

#### Created by belg4mit@mit.edu

This is a bug report for perl from belg4mit@&#8203;mit\.edu\,
generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
While playing with code to dup DATA to STDIN I inadvertently came
up with the following

&nbsp;  $a = tell\(DATA\);
&nbsp;  seek\(DATA\,$a\,0\);
&nbsp;  open\(STDIN\, "\<&DATA"\);
&nbsp;  while\(\<STDIN>\)\{
&nbsp;    print "STDIN&#8203;: $\_";
&nbsp;  \}
&nbsp;  \_\_DATA\_\_
&nbsp;  Foo
&nbsp;  Bar
&nbsp;  Quux

The first two lines would seem to be no\-ops\. And yet in several 
instances nothing is read without them\. Known configs where this
has failed are this box with the version of perl below as well
as the following perl \-V I have also received reports of failure
without the first two lines on builds of 5\.8 for SCO and RH

Summary of my perl5 \(revision 5\.0 version 6 subversion 1\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.8\, archname=sun4\-solaris
&nbsp;    uname='sunos downy\.mit\.edu 5\.8 generic\_108528\-12 sun4u sparc sunw\,ultraax\-i2 '
&nbsp;    config\_args='\-d \-s \-e \-Dprefix=/usr/athena \-Dcc=gcc \-Dlocincpth=/usr/athena/include \-Dloclibpth=/usr/athena/lib \-Dperladmin=bugs@&#8203;mit\.edu'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=undef d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/athena/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/athena/include'
&nbsp;    ccversion=''\, gccversion='3\.0\.3'\, gccosandvers='solaris2\.8'
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/athena/lib '
&nbsp;    libpth=/usr/athena/lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-lgdbm \-ldb \-ldl \-lm \-lc \-lbsd \-liconv
&nbsp;    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lc \-lbsd \-liconv
&nbsp;    libc=/lib/libc\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' \-Wl\,\-E'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags=' \-W\,l\-E \-G \-L/usr/athena/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under solaris
&nbsp;  Compiled at Sep 12 2002 16&#8203;:20&#8203;:01
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /usr/athena/lib/perl5/5\.6\.1/sun4\-solaris
&nbsp;    /usr/athena/lib/perl5/5\.6\.1
&nbsp;    /usr/athena/lib/perl5/site\_perl/5\.6\.1/sun4\-solaris
&nbsp;    /usr/athena/lib/perl5/site\_perl/5\.6\.1
&nbsp;    /usr/athena/lib/perl5/site\_perl
&nbsp;    \.
Summary of my perl5 \(5\.0 patchlevel 4 subversion 4\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.5\.1\, archname=sun4\-solaris
&nbsp;    uname='sunos technomage 5\.5\.1 generic\_103640\-12 sun4m sparc sunw\,sparcclassic '
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    bincompat3=y useperlio=undef d\_sfio=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='/usr/gcc/bin/gcc'\, optimize='\-O2'\, gccversion=2\.7\.2\.1
&nbsp;    cppflags=''
&nbsp;    ccflags =''
&nbsp;    stdchar='unsigned char'\, d\_stdstdio=define\, usevfork=false
&nbsp;    voidflags=15\, castflags=0\, d\_casti32=define\, d\_castneg=define
&nbsp;    intsize=4\, alignbytes=8\, usemymalloc=y\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='/usr/gcc/bin/gcc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-ldl \-lm \-lc \-lcrypt
&nbsp;    libc=/lib/libc\.so\, so=so
&nbsp;    useshrplib=false\, libperl=libperl\.a
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-G \-L/usr/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Built under solaris
&nbsp;  Compiled at Feb  9 1998 20&#8203;:13&#8203;:09
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /afs/athena/contrib/perl5/arch/sun4x\_55/lib/perl5\.004
&nbsp;    /afs/athena/contrib/perl5/arch/share/perl5\.004
&nbsp;    /afs/athena/contrib/perl5/arch/sun4x\_55/lib/perl5\.004/site\_perl
&nbsp;    /afs/athena/contrib/perl5/arch/share/perl5\.004/site\_perl
&nbsp;    \.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.0:

Configured by belg4mit at Fri Aug  9 06:33:44 EDT 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=solaris, osvers=2.8, archname=sun4-solaris-64int
    uname='sunos biohazard-cafe.mit.edu 5.8 generic_108528-12 sun4u sparc sunw,sun-fire-280r '
    config_args='-Dprefix=/mit/belg4mit'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=undef usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=undef
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='',
    optimize='-O',
    cppflags=''
    ccversion='Sun WorkShop 6 update 1 C 5.2 2000/09/11', gccversion='', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=87654321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=4
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/mit/sunsoft_v6.1a/SUNWspro/WS6U1/lib'
    libpth=/usr/local/lib /mit/sunsoft_v6.1a/SUNWspro/WS6U1/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags='-G -L/usr/local/lib -L/mit/sunsoft_v6.1a/SUNWspro/WS6U1/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /mit/belg4mit/lib/perl5/5.8.0/sun4-solaris-64int
    /mit/belg4mit/lib/perl5/5.8.0
    /mit/belg4mit/lib/perl5/site_perl/5.8.0/sun4-solaris-64int
    /mit/belg4mit/lib/perl5/site_perl/5.8.0
    /mit/belg4mit/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/mit/belg4mit
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/usr/openwin/lib
    LOGDIR (unset)
    PATH=/mit/belg4mit/Private:/mit/belg4mit/arch/sun4x_58/bin:/srvd/patch:/usr/athena/bin:/usr/athena/etc:/bin/athena:/usr/openwin/bin:/usr/openwin/demo:/usr/bin:/usr/ccs/bin:/usr/sbin:/sbin:/usr/ucb:/mit/sipb/arch/sun4x_58/bin:/mit/gnu/arch/sun4x_58/bin:/mit/games/arch/sun4x_58/bin:/mit/consult/arch/sun4x_58/bin:/mit/outland/arch/sun4x_58/bin:/mit/perl5/arch/sun4x_58/bin:/mit/perl/arch/sun4x_58/bin
    PERL_BADLANG (unset)
    SHELL=/bin/athena/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Strange how these things go unnoticed for ages then get discovered 
by several people at nearly same time\.
This is caused by "fact" that DATA handle has been read in order to 
read the script\. So some data is in the buffer\. The "dup" process 
does not copy the buffer\.

Not clear if/how this should be fixed&#8203;:
&nbsp;  A\. Document it
&nbsp;  B\. Make dup do seek\(FH\,tell\(FH\)\,0\)  \- but with more error checking\.
&nbsp;  C\. Make DATA file handle do the seek/tell thing implicitly\.

**Comment by p5pRT** [comment]
nick@ing-simmons.net - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Mar 31 18&#8203;:43&#8203;:53 2003\, belg4mit@&#8203;mit\.edu wrote&#8203;:
> This is a bug report for perl from belg4mit@&#8203;mit\.edu\,
> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> While playing with code to dup DATA to STDIN I inadvertently came
> up with the following
> 
>   $a = tell\(DATA\);
>   seek\(DATA\,$a\,0\);
>   open\(STDIN\, "\<&DATA"\);
>   while\(\<STDIN>\)\{
>     print "STDIN&#8203;: $\_";
>   \}
>   \_\_DATA\_\_
>   Foo
>   Bar
>   Quux
> 
> The first two lines would seem to be no\-ops\. And yet in several
> instances nothing is read without them\. Known configs where this
> has failed are this box with the version of perl below as well
> as the following perl \-V \[all Solaris\.\] I have also received reports
of failure
> without the first two lines on builds of 5\.8 for SCO and RH
> 

On Sat May 10 10&#8203;:32&#8203;:12 2003\, ni\-s wrote&#8203;:
> Strange how these things go unnoticed for ages then get discovered 
> by several people at nearly same time\.
> This is caused by "fact" that DATA handle has been read in order to 
> read the script\. So some data is in the buffer\. The "dup" process 
> does not copy the buffer\.
> 
> Not clear if/how this should be fixed&#8203;:
>   A\. Document it
>   B\. Make dup do seek\(FH\,tell\(FH\)\,0\)  \- but with more error checking\.
>   C\. Make DATA file handle do the seek/tell thing implicitly\.

ni\-s is no longer with us to handle this old ticket\.

This problem appears to be OS\-specific; I could not reproduce it on
Darwin or Debian Linux\.  Is there someone \-\- especially on Solaris \-\-
who can try to reproduce this bug?  How should we approach this?

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @perhunter

On 04/21/2012 12&#8203;:02 PM\, James E Keenan via RT wrote&#8203;:
> On Mon Mar 31 18&#8203;:43&#8203;:53 2003\, belg4mit@&#8203;mit\.edu wrote&#8203;:
>> This is a bug report for perl from belg4mit@&#8203;mit\.edu\,
>> generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
>>
>>
>> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
>> While playing with code to dup DATA to STDIN I inadvertently came
>> up with the following
>>
>>    $a = tell\(DATA\);
>>    seek\(DATA\,$a\,0\);
>>    open\(STDIN\, "\<&DATA"\);
>>    while\(\<STDIN>\)\{
>>      print "STDIN&#8203;: $\_";
>>    \}
>>    \_\_DATA\_\_
>>    Foo
>>    Bar
>>    Quux
>>
>> The first two lines would seem to be no\-ops\. And yet in several
>> instances nothing is read without them\. Known configs where this
>> has failed are this box with the version of perl below as well
>> as the following perl \-V \[all Solaris\.\] I have also received reports
> of failure
>> without the first two lines on builds of 5\.8 for SCO and RH
>>
>
> On Sat May 10 10&#8203;:32&#8203;:12 2003\, ni\-s wrote&#8203;:
>> Strange how these things go unnoticed for ages then get discovered
>> by several people at nearly same time\.
>> This is caused by "fact" that DATA handle has been read in order to
>> read the script\. So some data is in the buffer\. The "dup" process
>> does not copy the buffer\.
>>
>> Not clear if/how this should be fixed&#8203;:
>>    A\. Document it
>>    B\. Make dup do seek\(FH\,tell\(FH\)\,0\)  \- but with more error checking\.
>>    C\. Make DATA file handle do the seek/tell thing implicitly\.
>
> ni\-s is no longer with us to handle this old ticket\.
>
> This problem appears to be OS\-specific; I could not reproduce it on
> Darwin or Debian Linux\.  Is there someone \-\- especially on Solaris \-\-
> who can try to reproduce this bug?  How should we approach this?

just fyi\, i ran into this problem without the dup call\. in File&#8203;::Slurp 
slurping the DATA handle which uses sysread has seek issues\. my 
workaround was to detect it was DATA and to to a sysseek/tell on the 
handle before reading it with sysread\. so it sounds like the same issue\. 
i recall it was solaris that had this problem but i can't check it as i 
don't have solaris around anymore\. so having dup do the seek isn't going 
to fix the problem as it shows up without dup\.

uri

**Comment by p5pRT** [comment]
### From @Hugmeir

On Sat\, Apr 21\, 2012 at 1&#8203;:02 PM\, James E Keenan via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

> On Mon Mar 31 18&#8203;:43&#8203;:53 2003\, belg4mit@&#8203;mit\.edu wrote&#8203;:
> > This is a bug report for perl from belg4mit@&#8203;mit\.edu\,
> > generated with the help of perlbug 1\.34 running under perl v5\.8\.0\.
> >
> >
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > While playing with code to dup DATA to STDIN I inadvertently came
> > up with the following
> >
> >   $a = tell\(DATA\);
> >   seek\(DATA\,$a\,0\);
> >   open\(STDIN\, "\<&DATA"\);
> >   while\(\<STDIN>\)\{
> >     print "STDIN&#8203;: $\_";
> >   \}
> >   \_\_DATA\_\_
> >   Foo
> >   Bar
> >   Quux
> >
> > The first two lines would seem to be no\-ops\. And yet in several
> > instances nothing is read without them\. Known configs where this
> > has failed are this box with the version of perl below as well
> > as the following perl \-V \[all Solaris\.\] I have also received reports
> of failure
> > without the first two lines on builds of 5\.8 for SCO and RH
> >
>
> On Sat May 10 10&#8203;:32&#8203;:12 2003\, ni\-s wrote&#8203;:
> > Strange how these things go unnoticed for ages then get discovered
> > by several people at nearly same time\.
> > This is caused by "fact" that DATA handle has been read in order to
> > read the script\. So some data is in the buffer\. The "dup" process
> > does not copy the buffer\.
> >
> > Not clear if/how this should be fixed&#8203;:
> >   A\. Document it
> >   B\. Make dup do seek\(FH\,tell\(FH\)\,0\)  \- but with more error checking\.
> >   C\. Make DATA file handle do the seek/tell thing implicitly\.
>
> ni\-s is no longer with us to handle this old ticket\.
>
> This problem appears to be OS\-specific; I could not reproduce it on
> Darwin or Debian Linux\.  Is there someone \-\- especially on Solaris \-\-
> who can try to reproduce this bug?  How should we approach this?
>
>

On Solaris 10\, I can't reproduce this on 5\.12\.4\, 5\.14\.2\, or blead\, but
could on the system\-provided 5\.8\.4; So this was fixed somewhere along the
road\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
